design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,201 @@
+
+This KLIP attempts to integrate producing test data into the KSQL REPL.
+
+## Scope
+
+This KLIP is intended to improve the *development and testing* workflow as opposed to providing
+*production-ready* functionality. The following tasks will be addressed:
+
+* KSQL will be able to create and produce data to arbitrary topics
+* KSQL will be able to validate data when producing to a topic that is registered with a schema
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO
+
+```sql
+INSERT INTO (stream_name|table_name) (column1, column2, ...)
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT INTO` statement can be used to insert new records into a kafka topic (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. The key for the insert statement is taken from the
+column the matches the name of `key` in the `CREATE` statement properties for the corresponding
+source. This functionality will need to be extended when we support structured keys.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. Analogously, the key field will be taken from the same value as specified in the
+`key` property. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (i.e. by following syntax 
+like Hive's `NAMED_STRCUT` construct). 
+
+Future extensions: 
+* Support some`AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+```sql
+CREATE TABLE table_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+The `WITH` clause will now accept two additional property_names: `PARTITIONS`/`REPLICAS` If the 
+topic *does not exist* and these values are specified, this command will create a topic with a the
+set number of partitions and replicas before executing the command. If the topic exists and has
+a different number of partitions/replicas, this command will fail.
+
+This functionality will be flagged under a configuration `ksql.create.source.allow.empty.topic`,
```

Comment:
How about `source.topic.create.allow` instead? I find the word `empty` confusing since to me `empty` means there are no data in the topic, not that the topic doesn't exist.

Reply:
I like your suggestion!

Reply:
I settled on `ksql.ddl.source.topic.create.allow` to differentiate it from `CSAS`, which already allows this behavior.